XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,419 @@
+four areas. On instantiation, before/after the callback is called and when the
+instance is destructed. For cases where resources are reused, instantiation and
+destructor calls are emulated.
+
+
+#### `init(id, type, parentId)`
+
+* `id` {Number}
+* `type` {String}
+* `parentId` {Number}
+
+Called when a class is constructed that has the possibility to trigger an
+asynchronous event. This does not mean the instance will trigger a
+`pre()`/`post()` event before `destroy()` is called. Only that the possibility
+exists.
+
+This behavior can be observed by doing something like opening a resource then
+closing it before the resource can be used. The following snippet demonstrates
+this.
+
+```js
+require('net').createServer().listen(8080, function() { this.close() });
+```
+
+Every instance is assigned a unique id. Taking advantage of the fraction space
+in the 64-bit IEEE 754 that ECMAScript defines as the **Number** type, the
+number of id's that can be assigned are `2^53 - 1` (also defined as
+`Number.MAX_SAFE_INTEGER`). At this size node can assign a new id every 100
+nanoseconds and not run out for over 28 years. Because of this circumstance it
+is not deemed necessary to find an alternative approach.
+
+`parentId` is the unique id of either the async resource at the top of the JS
+stack when `init()` was called, or the originator of the new resource. For
+example, when a connection is made to a TCP server there is no JS stack
+available when the connection's constructor is executed. Meaning there would be
+no `parentId` available on the JS stack. Instead the TCP server's unique id is
+manually passed to the client's constructor and propagated to the `init()`'s
+`parentId`.
+
+
+#### `pre(id, entryPoint)`
+
+* `id` {Number}
+* `entryPoint` {Function}
+
+Called just before the return callback is called after completing an
+asynchronous request. Or called on handles with events such as receiving a new
+connection. For requests, such as `fs.open()`, this should be called exactly
+once. For handles, such as a TCP server, this may be called 0-N times.
+
+The `entryPoint` is the callback that will be executed immediately after
+`pre()` returns. This is being passed as a way to track what work will be done.
+
+**Note(trevnorris):** I'm not sure passing only `entryPoint` is useful. Since
+the arguments of the function to be called will be available it may be useful
+to pass those to `pre()` as well. Though I have reservations about the user
+being able to mess with non-primitives.
+
+
+#### `post(id, didThrow)`
+
+* `id` {Number}
+* `didThrow` {Boolean}
+
+Called immediately after the return callback is completed. If the callback
+threw but was caught by a domain or `uncaughtException`, `didThrow` will be set
+to `true`. If the callback threw but was not caught then the process will exit
+immediately without calling `post()`.
+
+
+#### `destroy(id)`
+
+* `id` {Number}
+
+Called either when the class destructor is run, or if the resource is marked as
```

Comment:
It's called during deconstruction of the C++ class. e.g. during the execution of `delete`. Or during the "conceptual" deconstruction of the resource if the resource is going to be placed in a queue for later use (e.g. `HTTPParser`).

Will clarify.